mypy/newsemanal/semanal.py

```diff
@@ -2118,6 +2118,8 @@ def unwrap_final(self, s: AssignmentStmt) -> bool:
             return False
         lval = s.lvalues[0]
         assert isinstance(lval, RefExpr)
+        # Reset inferred status if it was set due to simple literal rvalue on previous iteration.
+        lval.is_inferred_def = s.type is None
```

Comment:
This seems error-prone and hard to reason about. For example, what happens if there are multiple assignments to a final name? How hard it would be to avoid the need to reset this (by not storing the bad value perhaps)?

Reply:
What do you mean by "bad" value? The value of `is_inferred_def` is set correctly, but just few lines above we reset `s.type` to `None`, but not the flag. If we don't reset it together, `s.type` will stay `None`.

> For example, what happens if there are multiple assignments to a final name?

There will be an inconsistency with old analyzer. But we anyway give an error. Also this can be fixed only doing the reset if `lvalue.is_new_def`.

Anyway, if you have a better fix, go ahead.